opencog/cython/PythonEval.cc

```diff
@@ -985,114 +985,76 @@ PyObject* PythonEval::call_user_function(const std::string& moduleFunction,
-    // Get the atom object returned by this user function.
-    BOOST_SCOPE_EXIT(void) {
-        pop_context_atomspace();
-    } BOOST_SCOPE_EXIT_END
-    PyObject* pyReturnAtom = this->call_user_function(func, varargs);
-
-    // If we got a non-null atom were no errors.
-    if (pyReturnAtom)
-    {
-        // Grab the GIL.
-        PyGILState_STATE gstate;
-        gstate = PyGILState_Ensure();
-
-        // Get the handle from the atom.
-        PyObject* pyAtomPATOM = PyObject_CallMethod(pyReturnAtom,
-                (char*) "handle_ptr", NULL);
-
-        // Make sure we got an atom pointer.
-        PyObject* pyError = PyErr_Occurred();
-        if (pyError or nullptr == pyAtomPATOM)
-        {
-            PyGILState_Release(gstate);
-            throw RuntimeException(TRACE_INFO,
-                "Python function '%s' did not return Atom!", func.c_str());
-        }
-
-        // Get the atom pointer from the python atom pointer.
-        // Save it, because the DECREF will blow it away.
-        Handle hresult = *((Handle*)(PyLong_AsLong(pyAtomPATOM)));
-
-        // Cleanup the reference counts.
-        Py_DECREF(pyReturnAtom);
-        Py_DECREF(pyAtomPATOM);
-
-        // Release the GIL. No Python API allowed beyond this point.
-        PyGILState_Release(gstate);
-        return hresult;
-    }
-    else
-    {
-        throw RuntimeException(TRACE_INFO,
-            "Python function '%s' did not return Atom!", func.c_str());
-    }
-
-    return Handle::UNDEFINED;
-}
-
 /**
  * Apply the user function to the arguments passed in varargs and
- * return the extracted truth value.
+ * return the extracted Value.
  */
-TruthValuePtr PythonEval::apply_tv(AtomSpace * as,
-                                   const std::string& func,
-                                   Handle varargs)
+ValuePtr PythonEval::apply_v(AtomSpace * as,
+                             const std::string& func,
+                             Handle varargs)
 {
     std::lock_guard<std::recursive_mutex> lck(_mtx);
     push_context_atomspace(as);
     BOOST_SCOPE_EXIT(void) {
         pop_context_atomspace();
     } BOOST_SCOPE_EXIT_END
-    // Get the python truth value object returned by this user function.
-    PyObject *pyTruthValue = call_user_function(func, varargs);
 
-    // If we got a non-null truth value there were no errors.
-    if (NULL == pyTruthValue)
+    // Get the python value object returned by this user function.
+    PyObject *pyValue = call_user_function(func, varargs);
+
+    // If we got a non-null Value there were no errors.
+    if (NULL == pyValue)
         throw RuntimeException(TRACE_INFO,
-            "Python function '%s' did not return TruthValue!",
+            "Python function '%s' did not return Atomese!",
             func.c_str());
 
     // Grab the GIL.
     PyGILState_STATE gstate = PyGILState_Ensure();
 
+    // Did we actually get a Value?
+    if (0 == PyObject_HasAttrString(pyValue, "value_ptr"))
```

Comment:
Checking if `pyValue` is subclass of the `Value` would be more safe. This is few more lines of code though.

Reply:
Other code looks good to me.

Reply:
pyValue isn't a subclass of Value, its a pointer to a python object holding a pointer to a value.

Reply:
Unless you mean that I should call `PyObject_IsInstance()` ...

Reply:
Yes, I mean `PyObject_IsInstance`